### PR TITLE
Json schema for Coin and some parts of Transfer

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
Closes: #772

## Description

We do coin router kind of thing in CosmWasm. It needs some IBC things, and it requests JsonSchema on types.
So I did it.

Also, I made serde of Amount consistently decimal(decimal string). It was from str decimal, and in coin it was decimal, but on its own if was hex string. So made on its own also decimal. For some it can be breaking change. But other way schema (and CosmWasm messages) would be bad/inconsistent.
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR. may be i guess @Farhad-Shabani 

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/signer.rs [hunk 1]
```diff
@@ -2,6 +2,9 @@ use crate::prelude::*;
 
 use derive_more::Display;
 
+#[cfg(feature = "schema")]
+use crate::alloc::borrow::ToOwned;
```
Comment (Farhad-Shabani) [inline crates/ibc/src/signer.rs]: Isn't this redundant?

File: .changelog/unreleased/improvement/772-json-schema-for-channels-and-coin.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- Json schema for Coin and some parts of Transfer
```
Comment (Farhad-Shabani) [inline .changelog/unreleased/improvement/772-json-schema-for-channels-and-coin.md]: Given the changes to `serde` on `Amount`, let's move this to the `breaking-changes`.

**Review by Farhad-Shabani — CHANGES_REQUESTED** [review]
> We do coin router kind of thing in CosmWasm. It needs some IBC things, and it requests JsonSchema on types.

Great seeing you're using ibc-rs! Thanks for featuring this :)

> Also, I made serde of Amount consistently decimal(decimal string). It was from str decimal, and in coin it was decimal, but on its own if was hex string. So made on its own also decimal. For some it can be breaking change. But other way schema (and CosmWasm messages) would be bad/inconsistent.

Just got a deserialization issue when tested locally. Otherwise, we are good.

- Before, serialization roundtrip (ser > deser) on both `Coin` and `Amount` was ending up with a genuine value. Now, on deserialization, it fails. I think we should add a tailored `deserialize` for `U256` instead of using [this](https://github.com/cosmos/ibc-rs/blob/98099399137b544e66d50af45ecfc8c8e508d0b2/crates/ibc/src/serializers.rs#L29-L38). 
- Also, It would be great if you include a test for it.

File: crates/ibc/src/applications/transfer/amount.rs [hunk 3]
```diff
@@ -6,10 +6,39 @@ use derive_more::{Display, From, Into};
...
+);
+
+#[cfg(feature = "serde")]
+fn deserialize<'de, D>(deserializer: D) -> Result<U256, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    use serde::Deserialize;
+    U256::from_dec_str(<&str>::deserialize(deserializer)?).map_err(serde::de::Error::custom)
+}
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/amount.rs]: Can you place this under the [`seriliazers.rs`](https://github.com/cosmos/ibc-rs/blob/main/crates/ibc/src/serializers.rs#L0-L1) module? Also, `#[cfg(feature = "serde")]` not needed as having it derived on `seriliazers`

File: crates/ibc/src/applications/transfer/amount.rs [hunk 4]
```diff
@@ -6,10 +6,39 @@ use derive_more::{Display, From, Into};
...
+#[test]
+#[cfg(feature = "serde")]
+fn serde_amount() {
+    let value = Amount::from(42);
+    let string = serde_json::to_string(&value).expect("can serde string");
+    assert_eq!(string, "\"42\"");
+    let binary = serde_json::to_vec(&value).expect("can serde binary");
+    let de: Amount = serde_json::from_slice(binary.as_ref()).expect("can deserialize");
+    assert_eq!(de, value);
+}
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/amount.rs]: And move this down to the bottom of the page under `mod tests {}?` like e.g. [here](https://github.com/cosmos/ibc-rs/blob/main/crates/ibc/src/clients/ics07_tendermint/consensus_state.rs#L162-L183)

**Review by Farhad-Shabani — CHANGES_REQUESTED** [review]
Awesome!

A quick nit and then done!

File: crates/ibc/src/applications/transfer/amount.rs (hunk 3 above)
Comment (dzmitry-lahoda) [reply crates/ibc/src/applications/transfer/amount.rs]: i am not sure about it. any use for U256 from dec string rather than amount?

**Review by Farhad-Shabani — APPROVED** [review]
:pray:

File: crates/ibc/src/applications/transfer/amount.rs (hunk 3 above)
Comment (dzmitry-lahoda) [reply crates/ibc/src/applications/transfer/amount.rs]: until it public, not use for this https://github.com/cosmos/ibc-rs/pull/791#issuecomment-1658313262

**Pull request merged**